### PR TITLE
[SMTChecker] Add ArraySort and array operations

### PR DESCRIPTION
**Pull request by leonardoalt**
Split #5387 

Introduces `ArraySort` and its operations `select` and `store`.

### CODE REVIEW

File: libsolidity/formal/SymbolicTypes.cpp [hunk 1]
```diff
@@ -43,6 +43,10 @@ smt::SortPointer dev::solidity::smtSort(Type const& _type)
 		smt::SortPointer returnSort = smtSort(*returnTypes.at(0));
 		return make_shared<smt::FunctionSort>(parameterSorts, returnSort);
 	}
+	case smt::Kind::Array:
+	{
+		solAssert(false, "Invalid type");
+	}
```
Comment (leonardoalt) [inline libsolidity/formal/SymbolicTypes.cpp:49]: Will be implemented in the PR that adds support to `mapping`.

File: libsolidity/formal/SolverInterface.h [hunk 2]
```diff
@@ -79,6 +80,18 @@ struct FunctionSort: public Sort
 	}
 };
 
+struct ArraySort: public Sort
+{
+	ArraySort(SortPointer _domain, SortPointer _range):
```
Comment (axic) [inline libsolidity/formal/SolverInterface.h:87]: Please leave a comment explaining the parameters. I suppose: domain is the type of the values and range is the type of the index?

File: libsolidity/formal/SolverInterface.h [hunk 3]
```diff
@@ -127,6 +142,33 @@ class Expression
 		return !std::move(_a) || std::move(_b);
 	}
 
+	static Expression select(Expression _array, Expression _index)
```
Comment (axic) [inline libsolidity/formal/SolverInterface.h:148]: Please leave a comment explaining what this is doing.

File: libsolidity/formal/SolverInterface.h [hunk 4]
```diff
@@ -127,6 +142,33 @@ class Expression
...
+		solAssert(arraySort, "");
+		solAssert(*arraySort->domain == *_index.sort, "");
+		return Expression(
+			"select",
+			std::vector<Expression>{std::move(_array), std::move(_index)},
+			arraySort->range
+		);
+	}
+
+	static Expression store(Expression _array, Expression _index, Expression _element)
```
Comment (axic) [inline libsolidity/formal/SolverInterface.h:163]: Please leave a comment explaining what this is doing.

File: libsolidity/formal/SymbolicTypes.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/formal/SymbolicTypes.cpp:49]: `solUnimplementedAssert` in that case

**Comment by leonardoalt** [comment]
@axic Updated

File: libsolidity/formal/SolverInterface.h [hunk 5]
```diff
@@ -79,6 +80,20 @@ struct FunctionSort: public Sort
 	}
 };
 
+struct ArraySort: public Sort
+{
+	/// _domain is the sort of the indices
+	/// _range is the type of the values
```
Comment (axic) [inline libsolidity/formal/SolverInterface.h]: Be consistent between sort/type :)

**Review by axic — APPROVED** [review]

**Pull request merged**